### PR TITLE
Teach ds2 about first-chance exceptions on Windows.

### PR DESCRIPTION
**Pull request by fischman-bcny**
Absent this, calling WindowsXamlManager.initializeForCurrentThread crashes the inferior as an (expected, and caught up-stack) WinRT exception is instead interpreted as a fatal error.

## Tested
In one terminal:
`& "<path-to>\ds2\out\ds2.exe" g :5800 -- <path/to/my-executable.exe> --my-args`
In another terminal:
`lldb -o 'gdb-remote 5800' -o 'c' -- <path/to/my-executable.exe> --my-args`

### CODE REVIEW

File: Sources/Target/Windows/Thread.cpp [hunk 1]
```diff
@@ -119,11 +124,24 @@ void Thread::updateState(DEBUG_EVENT const &de) {
...
     DS2LOG(
-        Debug, "exception from inferior, tid=%lu, code=%s, address=%" PRI_PTR,
+        Debug, "exception from inferior, tid=%lu, code=%s, dwFirstChance=%lu, address=%" PRI_PTR,
         tid(),
         Stringify::ExceptionCode(de.u.Exception.ExceptionRecord.ExceptionCode),
+        de.u.Exception.dwFirstChance,
         PRI_PTR_CAST(de.u.Exception.ExceptionRecord.ExceptionAddress));
 
+      // First-chance exceptions may yet be caught by the debugee, so don't break for them.
+      // An exception is STATUS_BREAKPOINT, since Process::initialize()'s implementation relies on
```
Comment (compnerd) [inline Sources/Target/Windows/Thread.cpp:134]: I'm afraid I don't understand this comment.  We do not control the `DebugEvent` right?  So, the `ExceptionCode` being `STATUS_BREAKPOINT` is defined by the system is my understanding.

**Review by compnerd — COMMENTED** [review]
I find the change somewhat confusing.  The model representing a thread in the inferior is being updated to match the current state and that causes the inferior to resume.  I think that we should hoist the process resumption to a higher level (caller) instead so that the update of the model in the debug server does not cause a side effect on the inferior.

File: Sources/Target/Windows/Thread.cpp (hunk 1 above)
Comment (fischman-bcny) [reply Sources/Target/Windows/Thread.cpp:134]: Yes that's my understanding as well. Not sure where the mismatch with the comment is in your eyes.

Comment (fischman-bcny) [reply Sources/Target/Windows/Thread.cpp]: Was able to narrow the condition and simplify the comment.

**Comment by fischman-bcny** [comment]
> I think that we should hoist the process resumption to a higher level

Done.
@compnerd PTAL.

File: Sources/Target/Windows/Process.cpp [hunk 2]
```diff
@@ -259,7 +259,21 @@ ErrorCode Process::wait() {
...
+      // First-chance exceptions may yet be caught by the debugee, so don't break for them.
+      auto code = de.u.Exception.ExceptionRecord.ExceptionCode;
+      if (de.u.Exception.dwFirstChance &&
+          (code == DS2_EXCEPTION_UNCAUGHT_COM || code == DS2_EXCEPTION_UNCAUGHT_USER || code == DS2_EXCEPTION_UNCAUGHT_WINRT)) {
+        DS2LOG(Error, "Ignoring first-chance exception and continuing; code: %s",
+               Stringify::ExceptionCode(de.u.Exception.ExceptionRecord.ExceptionCode));
+        BOOL result = ::ContinueDebugEvent(pid(), de.dwThreadId, DBG_EXCEPTION_NOT_HANDLED);
+        if (!result) {
+          return Platform::TranslateError();
+        }
```
Comment (compnerd) [inline Sources/Target/Windows/Process.cpp:272]: I'd just collapse this:

```suggestion
        if (!::ContinueDebugEvent(pid(), de.dwThreadId, DBG_EXCEPTION_NOT_HANDLED))
          return Platform::TranslateError();
```

File: Sources/Target/Windows/Process.cpp [hunk 3]
```diff
@@ -259,7 +259,21 @@ ErrorCode Process::wait() {
       continue;
     }
 
-    case EXCEPTION_DEBUG_EVENT:
+    case EXCEPTION_DEBUG_EVENT: {
+      // First-chance exceptions may yet be caught by the debugee, so don't break for them.
+      auto code = de.u.Exception.ExceptionRecord.ExceptionCode;
+      if (de.u.Exception.dwFirstChance &&
+          (code == DS2_EXCEPTION_UNCAUGHT_COM || code == DS2_EXCEPTION_UNCAUGHT_USER || code == DS2_EXCEPTION_UNCAUGHT_WINRT)) {
```
Comment (compnerd) [inline Sources/Target/Windows/Process.cpp:266]: Should we be filtering the code or should we be ignoring all first-chance exceptions?

Comment (fischman-bcny) [reply Sources/Target/Windows/Process.cpp:266]: Doing the latter doesn't work, b/c e.g. apparently some first-chance exceptions (looking at you, DS2_EXCEPTION_VC_THREAD_NAME_SET) need to be continued with DBG_CONTINUE. 
I started down that path before deciding that the rest of the roster of codes is different and so this is the least unpalatable option IMO.

File: Sources/Target/Windows/Process.cpp (hunk 2 above)
Comment (fischman-bcny) [reply Sources/Target/Windows/Process.cpp]: Done. This is my preference as well, but I was following the existing art in these files which seems to pull out a temporary.
(I believe some people this makes code more debuggable but IMO that optimizes for the wrong thing)

File: Sources/Target/Windows/Process.cpp (hunk 3 above)
Comment (fischman-bcny) [reply Sources/Target/Windows/Process.cpp]: Updated per tuple.

**Review by compnerd — APPROVED** [review]

**Pull request merged**